### PR TITLE
Adds `amp-toolbox-cache-url`

### PR DESCRIPTION
**Pull request by andreban**
The cache-url component will be used by update-ping and cache-update APIs.

### CODE REVIEW

File: cache-url/README.md [hunk 1]
```diff
@@ -0,0 +1,11 @@
+# AMP-Toolbox Cache URL
+
+Translates an url from the origin to the AMP Cache URL format, according to the specification
```
Comment (sebastianbenz) [inline cache-url/README.md]: nit: s/url/URL

File: cache-url/README.md [hunk 2]
```diff
@@ -0,0 +1,11 @@
...
+available in the [AMP documentation](https://developers.google.com/amp/cache/overview).
+
+## Usage
+
+``javascript
+const ampUrl = require('amp-toolbox-cache-url');
+const cacheUrl = ampUrl.cacheUrl('cdn.ampproject.org', 'https://www.example.com');
```
Comment (sebastianbenz) [inline cache-url/README.md]: Have you considered simply returning a function? 

```
const createCacheUrl = require('amp-toolbox-cache-url');
createCacheUrl('cdn.ampproject.org', 'https://www.example.com');
```

File: cache-url/lib/AmpUrl.js [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+
+/**
+ * Translates canonical URLs into AMP Cache URLs, according to the specification available at
+ * https://developers.google.com/amp/cache/overview.
+ */
+class AmpUrl {
+
+  /**
+   * Translates the canonicalUrl to the AMP Cache equivalent, for a given AMP Cache.
+   * @param {string} updateCacheApiDomainSuffix the AMP Cache domain suffix
```
Comment (sebastianbenz) [inline cache-url/lib/AmpUrl.js]: suggestion: simplify to s/updateCacheApiDomainSuffix/domainSuffix

File: cache-url/lib/AmpUrl.js [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+   * @param {string} canonicalUrl the canonical URL
```
Comment (sebastianbenz) [inline cache-url/lib/AmpUrl.js]: Wyhy is this named canonical URL? suggestion: url

File: cache-url/lib/AmpUrl.js [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+
+    url.protocol = 'https';
+    url.hostname = punycode.toASCII(unicodeHostname) + '.' + updateCacheApiDomainSuffix;
+    url.pathname = pathSegment + originalHostname + url.pathname;
+    return url.toString();
+  }
+
+  /**
+   * Returns the AMP Cache path, based on the mime type of the file that is being loaded.
+   * @param {string} pathname the pathname on the canonical url.
```
Comment (sebastianbenz) [inline cache-url/lib/AmpUrl.js]: s/canonical// same as above

File: cache-url/lib/AmpUrl.js [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+ * https://developers.google.com/amp/cache/overview.
+ */
+class AmpUrl {
+
+  /**
+   * Translates the canonicalUrl to the AMP Cache equivalent, for a given AMP Cache.
+   * @param {string} updateCacheApiDomainSuffix the AMP Cache domain suffix
+   * @param {string} canonicalUrl the canonical URL
+   */
+  cacheUrl(updateCacheApiDomainSuffix, canonicalUrl) {
```
Comment (sebastianbenz) [inline cache-url/lib/AmpUrl.js]: s/updateCacheApiDomainSuffix/domainSuffix

File: cache-url/lib/AmpUrl.js [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+
+const {URL} = require('url');
+const punycode = require('punycode');
+const mime = require('mime-types');
+
+/**
+ * Translates canonical URLs into AMP Cache URLs, according to the specification available at
+ * https://developers.google.com/amp/cache/overview.
+ */
+class AmpUrl {
```
Comment (sebastianbenz) [inline cache-url/lib/AmpUrl.js]: Can you find a better name than AmpUrl? e.g. AmpCacheUrlCalculator

File: cache-url/spec/AmpUrlSpec.js [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+
+describe('AmpUrl', () => {
+  const ampUrl = new AmpUrl();
+  const cache = {
+    updateCacheApiDomainSuffix: 'cdn.ampproject.org'
+  };
+
+  describe('cacheUrl', () => {
+    const tests = [
+      {canonical: 'https://www.example.com',
```
Comment (sebastianbenz) [inline cache-url/spec/AmpUrlSpec.js]: Can you fix the formatting to make this easier to read? 

```
{
  canonical: 'http://www.example.com/index.html',
  cache: 'https://www-example-com.cdn.ampproject.org/c/www.example.com/index.html'
},
```

Comment (sebastianbenz) [reply cache-url/spec/AmpUrlSpec.js]: Also, I don't think we should name this url instead of canonical.

File: cache-url/spec/AmpUrlSpec.js [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+ * limitations under the License.
+ */
+
+'use strict';
+
+const AmpUrl = require('../lib/AmpUrl');
+
+describe('AmpUrl', () => {
+  const ampUrl = new AmpUrl();
+  const cache = {
```
Comment (sebastianbenz) [inline cache-url/spec/AmpUrlSpec.js]: Why is this a nested object instead of a simple string variable?

File: cache-url/package.json [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+    "url": "https://github.com/ampproject/amp-toolbox/issues"
+  },
+  "homepage": "https://github.com/ampproject/amp-toolbox#readme",
+  "devDependencies": {
+    "eslint": "^3.0.1",
+    "eslint-config-google": "^0.6.0",
+    "jasmine": "^2.9.0"
+  },
+  "dependencies": {
+    "amp-toolbox-cache-list": "<1.0.0",
```
Comment (sebastianbenz) [inline cache-url/package.json]: Is this dependency needed?

File: cache-url/spec/AmpUrlSpec.js [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+        cache: 'https://www-example-com.cdn.ampproject.org/c/www.example.com/index.html'},
+      {canonical: 'https://www.example.com/image.png',
+        cache: 'https://www-example-com.cdn.ampproject.org/i/s/www.example.com/image.png'},
+      {canonical: 'http://www.example.com/image.png',
+        cache: 'https://www-example-com.cdn.ampproject.org/i/www.example.com/image.png'},
+      {canonical: 'https://www.example.com/font.woff2',
+        cache: 'https://www-example-com.cdn.ampproject.org/r/s/www.example.com/font.woff2'},
+      {canonical: 'http://www.example.com/font.woff2',
+        cache: 'https://www-example-com.cdn.ampproject.org/r/www.example.com/font.woff2'}
+    ];
```
Comment (sebastianbenz) [inline cache-url/spec/AmpUrlSpec.js]: Can you add tests for different encodings and other corner cases?

File: cache-url/spec/AmpUrlSpec.js (hunk 8 above)
Comment (sebastianbenz) [reply cache-url/spec/AmpUrlSpec.js]: I think I would've simplified this to:

```
{
  'http://www.example.com/index.html': 'https://www-example-com.cdn.ampproject.org/c/www.example.com/index.html',
'http://www.example.com/index.html': 'https://www-example-com.cdn.ampproject.org/c/www.example.com/index.html',
...
},

File: cache-url/lib/AmpUrl.js (hunk 4 above)
Comment (andreban) [reply cache-url/lib/AmpUrl.js]: done.

File: cache-url/lib/AmpUrl.js (hunk 3 above)
Comment (andreban) [reply cache-url/lib/AmpUrl.js]: done.

File: cache-url/README.md (hunk 2 above)
Comment (andreban) [reply cache-url/README.md]: done.

File: cache-url/spec/AmpUrlSpec.js (hunk 8 above)
Comment (andreban) [reply cache-url/spec/AmpUrlSpec.js]: Fixed the formatting. I'm keeping the object, as it makes easier to use later in the code.

File: cache-url/spec/AmpUrlSpec.js (hunk 11 above)
Comment (andreban) [reply cache-url/spec/AmpUrlSpec.js]: done.

File: cache-url/package.json (hunk 10 above)
Comment (andreban) [reply cache-url/package.json]: removed.

File: cache-url/spec/AmpUrlSpec.js (hunk 9 above)
Comment (andreban) [reply cache-url/spec/AmpUrlSpec.js]: fixed.

**Comment by andreban** [comment]
@sebastianbenz I think all notes are fixed (it's been sitting here for a while, but I forgot to answer some comments). PTAL.

File: cache-url/lib/AmpCacheUrlGenerator.js [hunk 12]
```diff
@@ -0,0 +1,67 @@
...
+const {URL} = require('url');
+const punycode = require('punycode');
+const mime = require('mime-types');
+
+/**
+ * Translates the canonicalUrl to the AMP Cache equivalent, for a given AMP Cache.
+ * @param {string} domainSuffix the AMP Cache domain suffix
+ * @param {string} url the canonical URL
+ */
+function createCacheUrl(domainSuffix, url) {
```
[the remark this answers is not in the document]
Comment (andreban) [reply cache-url/lib/AmpCacheUrlGenerator.js]: Ideally, developers would get the suffix from the caches endpoint. Adding the default would mean hardcoding it.
